pkg/gadgets/run/tracer/tracer.go

```diff
@@ -236,6 +253,15 @@ func (t *Tracer) installTracer() error {
 				return fmt.Errorf("attach BPF program %q: %w", progName, err)
 			}
 			t.links = append(t.links, l)
+		} else if p.Type == ebpf.SocketFilter && strings.HasPrefix(p.SectionName, "socket") {
+			if t.useAttacherInterface {
+				return fmt.Errorf("several socket filters found, only one is supported")
+			}
+			t.useAttacherInterface = true
+			err := t.Tracer.AttachProg(t.collection.Programs[progName])
+			if err != nil {
+				return fmt.Errorf("attaching ebpf program to dispatcher: %w", err)
+			}
```

Comment:
With my comment above making a networktracer.Tracer a member, do you think we could have multiple instances of it in order to support multiple socket programs? (We can handle this later on)

Reply:
In theory yes, but with the current way the code is organised, it is not easy:

`AttachContainer()`/`DetachContainer()` on the `run` gadget can be called by the runtime as soon as `NewInstance()` / `PreGadgetRun()` is called, that is before `Run()` is called. But the `run` gadget only know the ebpf code when `Run()` method is called. So `AttachContainer()` should be made to work without knowing how many socket filters the program has.

This also means that the networkTracer's dispatcher ebpf program is installed even if the gadget does not have any socket filter. This is unfortunate.

I was considering adding a container cache in the `run` gadget. But I think that would be at the wrong layer. Alternatively, the following initialization could be done after calling the `Run()` method:
* https://github.com/inspektor-gadget/inspektor-gadget/blob/3653cd35afc97383ed49cbef0b4b065989a5faaa/pkg/operators/localmanager/localmanager.go#L322-L345
* https://github.com/inspektor-gadget/inspektor-gadget/blob/3653cd35afc97383ed49cbef0b4b065989a5faaa/pkg/operators/kubemanager/kubemanager.go#L283-L300

I suggest to handle this issue later.

Reply:
I agree, let's do this later on. Thanks for the clarification.